modules/casper.js

```diff
@@ -2515,6 +2515,79 @@ Casper.prototype.waitWhileVisible = function waitWhileVisible(selector, then, on
 };
 
 /**
+ * Waits until an program is executed (kills it if timeout), 
+ * to process a next step.
+ * Take care with programs that calls another processes, as it seems to kill only the child.pid
+ *
+ * @param  Array    command   array of parameters to be send to default system shell
+ * @param  Object   command  {program: program to execute, parameters: array of parameters to be send to program}
+ * @param  Function  then       The next step to perform (optional)
+ * @param  Function  onTimeout  A callback function to call on timeout (optional)
+ * @param  Number    timeout    The max amount of time to wait, in milliseconds (optional)
+ * @return Casper
+ */
+
+Casper.prototype.waitForExec = function waitForExec(command, then, onTimeout, timeout) {
+    "use strict";
+    timeout = getTimeoutAndCheckNextStepFunction(timeout, then, 'waitForExec', this.options.waitTimeout);
+    if (utils.isArray(command)) {
+        var system = require('system');
+        command = {program: (system.env.SHELL || system.env.ComSpec), parameters: command}; // SHELL for UNIX(?), ComSpec for Windows(?)
+    };
+    var fs = require('fs');
+    if ( ((!utils.isString(command.program)) || (!utils.isArray(command.parameters))) ) {
+        throw new CasperError("waitForExec() needs an object {program: string of program to execute, parameters: array of program parameters} or an array of program parameters to default system shell");
+    } else if (!fs.isExecutable(command.program)) {
+        this.log('Casper.waitForExec() is going to call non executable file ' + JSON.stringify(command.program) + ' - maybe runs if is in PATH', "warning");
+    };
+
+    var spawn = require("child_process").spawn;
+    var stdout = ''; // VARIABLE TO STORE PROGRAM STDOUT
+    var stderr = ''; // VARIABLE TO STORE PROGRAM STDERR
+    var exitCode = null; // VARIABLE TO STORE PROGRAM EXIT CODE
+    
+    var child = spawn(command.program, command.parameters);
+    
+    child.stdout.on("data", function (standardOut) { // keeps stdout updated
+        stdout += standardOut;
+    });
+    child.stderr.on("data", function (standardError) { // keeps stderr updated
+        stderr += standardError;
+    });
+    child.on("exit", function (code) {
+        exitCode = code;
+    });
+
+    function __details() {
+        return {data: {command: command, pid: child.pid, stdout: stdout, stderr: stderr, exitCode: exitCode}};
+    };
+    function __onTimeout(timeout, details) {
+        var __onWaitTimeout = onTimeout ? onTimeout : this.options.onWaitTimeout;
+        var signalToKill = "SIGTERM";
+        child.kill(signalToKill);
+        (function killAndCallOnWaitTimeout() {
```

Comment:
What do you think? Should we use `wait()` or `waitFor()` here?
What about add a new param killTimeout (so the method will be `waitForExec(command, then, onTimeout, timeout, killTimeout)`), for waiting `killTimeout` ms after SIGTERM and the SIGKILL to the child, for a more "polite" exit?

So it could be (with `waitFor()`):
```
killTimeout = ~~killTimeout || ~~this.options.waitTimeout;
// killTimeout = getTimeoutAndCheckNextStepFunction(killTimeout, then, 'waitForExec', this.options.waitTimeout, false);
// killTimeout = getTimeoutAndCheckNextStepFunction(killTimeout, then, 'killAndCallOnWaitTimeout', this.options.waitTimeout, false);
(function killAndCallOnWaitTimeout() {
	// I don't know if it should return
	// return this.waitFor(function isProgramKilled() { // HAVE TO ADD waitFor() TO MAKE child.on("exit"... UPDATES exitCode AND TO child.pid BE UPDATED
	this.waitFor(function isProgramKilled() { // HAVE TO ADD waitFor() TO MAKE child.on("exit"... UPDATES exitCode AND TO child.pid BE UPDATED
		return !child.pid;
	}, function onProgramKilled() { 
			this.log(f("waitForExec() has killed %s (PID %d) with %s", details.data.command.program, details.data.pid, signalToKill), "info");
			// this.then(this.createStep(__onWaitTimeout, timeout, __details()));
			__onWaitTimeout.call(this, timeout, __details());
	}, function onProgramNotKilled() {
			this.log(f("waitForExec() has not killed %s (PID %d) with %s", details.data.command.program, details.data.pid, signalToKill), "warning");
			signalToKill = (require('system').os.name !== "windows") ? "SIGKILL" : "WM_QUIT"; // "WM_QUIT" SEEMS TO BE THE WINDOWS EQUIVALENT TO UNIX SIGKILL
			// Change killTimeout to 0, because the process must be killed after SIGKILL, I don't think it makes sense to let it be another value
			killTimeout = 0
			child.kill(signalToKill);
			killAndCallOnWaitTimeout.call(this);
	}, killTimeout);
}).call(this);
```
Or (with `wait()`):
```
killTimeout = ~~killTimeout || ~~this.options.waitTimeout;
// killTimeout = getTimeoutAndCheckNextStepFunction(killTimeout, then, 'waitForExec', this.options.waitTimeout, false);
// killTimeout = getTimeoutAndCheckNextStepFunction(killTimeout, then, 'killAndCallOnWaitTimeout', this.options.waitTimeout, false);
(function killAndCallOnWaitTimeout() {
    // I don't know if it should return
	// return this.wait(killTimeout, function() { // HAVE TO ADD wait() TO MAKE child.on("exit"... UPDATES exitCode AND TO child.pid BE UPDATED
	this.wait(killTimeout, function() { // HAVE TO ADD wait() TO MAKE child.on("exit"... UPDATES exitCode AND TO child.pid BE UPDATED
		if (!child.pid) {
			this.log(f("waitForExec() has killed %s (PID %d) with %s", details.data.command.program, details.data.pid, signalToKill), "info");
			__onWaitTimeout.call(this, timeout, __details());
		} else {
			this.log(f("waitForExec() has not killed %s (PID %d) with %s", details.data.command.program, details.data.pid, signalToKill), "warning");
			signalToKill = (require('system').os.name !== "windows") ? "SIGKILL" : "WM_QUIT"; // "WM_QUIT" SEEMS TO BE THE WINDOWS EQUIVALENT TO UNIX SIGKILL
			// Change killTimeout to 0, because the process must be killed after SIGKILL, I don't think it makes sense to let it be another value
			killTimeout = 0
			child.kill(signalToKill);
			killAndCallOnWaitTimeout.call(this);
		};
	});
}).call(this);
```

I've tested both with the tests, and both worked.
If we add the `killTimeout`, remember to change the doc.